adapter/internal/sourcewatcher/sourcewatcher.go

```diff
@@ -0,0 +1,323 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package sourcewatcher
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/wso2/product-microgateway/adapter/config"
+	"github.com/wso2/product-microgateway/adapter/internal/api"
+	xds "github.com/wso2/product-microgateway/adapter/internal/discovery/xds"
+	"github.com/wso2/product-microgateway/adapter/internal/loggers"
+	"github.com/wso2/product-microgateway/adapter/internal/oasparser/model"
+	"github.com/wso2/product-microgateway/adapter/pkg/auth"
+	"github.com/wso2/product-microgateway/adapter/pkg/logging"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+)
+
+const (
+	apisArtifactDir string = "apis"
+	zipExt          string = ".zip"
+	branchHead      string = "refs/heads/"
+)
+
+var artifactsMap map[string]model.ProjectAPI
+
+// Start fetches the API artifacts at the startup and polls for changes from the remote repository
+func Start() {
+	conf, _ := config.ReadConfigs()
+
+	retryInterval := conf.Adapter.SourceControl.RetryInterval
+
+	loggers.LoggerSourceWatcher.Info("Starting source watcher")
+	// Fetch the API artifacts at the startup
+	repository, err := fetchArtifacts()
+
+	// Retry fetching the API artifacts if the first attempt fails
+	for {
+		if err == nil {
+			break
+		} else {
+			time.Sleep(time.Duration(retryInterval) * time.Second)
+			loggers.LoggerSourceWatcher.Info("Retrying fetching artifacts from the remote repository")
+			repository, err = fetchArtifacts()
+		}
+	}
+
+	artifactsMap, err = api.ProcessMountedAPIProjects()
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Processing API artifacts failed : %s", err.Error()),
+			Severity: logging.CRITICAL,
+			ErrorCode: 2500,
+		})
+	}
+
+	loggers.LoggerSourceWatcher.Info("Polling for changes")
+	go pollChanges(repository)
+}
+
+// fetchArtifacts clones the API artifacts from the remote repository into the artifacts directory in the adapter
+func fetchArtifacts() (repository *git.Repository,err error) {
+	conf, _ := config.ReadConfigs()
+
+	artifactsDirName := filepath.FromSlash(conf.Adapter.SourceControl.ArtifactsDirectory + "/" + apisArtifactDir)
+	repositoryURL := conf.Adapter.SourceControl.Repository.URL
+	branch := conf.Adapter.SourceControl.Repository.Branch
+
+	// Opens the local repository, if exists
+	repository, _ = git.PlainOpen(artifactsDirName)
+
+	// If a local repository exists, pull the changes from the remote repository
+	if repository != nil {
+		loggers.LoggerSourceWatcher.Info("Starting to fetch changes from remote repository")
+
+		compareRepository(repository)
```

Comment:
Shall we update the method name give the idea that it also pulls the changes? Maybe something like `pullRepositoryIfUpdated`